src/Features/Core/Portable/Completion/Providers/Snippets/AbstractSnippetCompletionProvider.cs

```diff
@@ -35,34 +47,115 @@ public override async Task<CompletionChange> GetChangeAsync(Document document, C
             var allTextChanges = await allChangesDocument.GetTextChangesAsync(document, cancellationToken).ConfigureAwait(false);
 
             var change = Utilities.Collapse(allChangesText, allTextChanges.AsImmutable());
-            return CompletionChange.Create(change, allTextChanges.AsImmutable(), newPosition: snippet.CursorPosition, includesCommitCharacter: true);
+            var finalTextChange = ExtendSnippetTextChange(change, snippet.Placeholders);
+            var lspSnippet = GenerateLSPSnippet(finalTextChange, snippet.Placeholders);
+            var props = ImmutableDictionary<string, string>.Empty
+                .Add("LSPSnippet", lspSnippet!);
+
+            return CompletionChange.Create(change, allTextChanges.AsImmutable(), properties: props, snippet.CursorPosition, includesCommitCharacter: true);
         }
 
-        public override async Task ProvideCompletionsAsync(CompletionContext context)
+        private static string? GenerateLSPSnippet(TextChange textChange, ImmutableArray<RoslynLSPSnippetItem> placeholders)
         {
-            var document = context.Document;
-            var cancellationToken = context.CancellationToken;
-            var position = context.Position;
-            var service = document.GetLanguageService<ISnippetService>();
+            var textChangeStart = textChange.Span.Start;
+            var textChangeText = textChange.NewText!;
+            var lspSnippetString = "";
 
-            if (service == null)
+            for (var i = 0; i < textChangeText.Length;)
             {
-                return;
+                var (str, strCount) = GetStringInPosition(placeholders, i, textChangeStart);
+                if (str.IsEmpty())
+                {
+                    lspSnippetString += textChangeText[i];
+                    i++;
+                }
+                else
+                {
+                    lspSnippetString += str;
+                    i += strCount;
+
+                    if (strCount == 0)
+                    {
+                        lspSnippetString += textChangeText[i];
+                        i++;
+                    }
+                }
             }
 
-            var (strippedDocument, newPosition) = await GetDocumentWithoutInvokingTextAsync(document, position, cancellationToken).ConfigureAwait(false);
+            return lspSnippetString;
+        }
+
+        private static (string, int) GetStringInPosition(ImmutableArray<RoslynLSPSnippetItem> placeholders, int position, int textChangeStart)
+        {
+            foreach (var placeholder in placeholders)
+            {
+                if (placeholder.CaretPosition.HasValue && placeholder.CaretPosition.Value - textChangeStart == position)
+                {
+                    return ("$0", 0);
```

Comment:
i don't relaly get how this works if the caret position is *prior* to the start of the textChange (or is that handled elsewhere?).